Simplify RID Model

richlander opened the pull request:
This plan is intended to:

- Freeze the RID graph.
- Adopt stable RIDs (no version numbers).
- Limit `runtimes.json` to being used by NuGet only (by default).
- Hosts adopt a simple non-extensible model.
- Contractualize what `linux-x64` means (same for Arm64 and x86).
- Define `linux-x64` vs source-build.

More generally, we waste significant amount of time talking about RIDs every release. This proposal is intended to simplify the most problematic aspect of the RID topic. Once that is resolved, we can think about taking on other RID challenges.

jkotas commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,221 @@
+
+The following are the biggest problems we see:
+
+- We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner], (https://github.com/dotnet/runtime/issues/65566)), with no end in sight.
```
```suggestion
- We get requests to add new RIDs ([Asianux Server 8](https://github.com/dotnet/runtime/issues/2129), [CBL-Mariner](https://github.com/dotnet/runtime/issues/65566), [Anolis Linux](https://github.com/dotnet/runtime/pull/66132)), with no end in sight.
```

yaakov-h commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,221 @@
+
+For .NET 8, we'll continue with the model. However, we will no longer be able to use [CentOS 7 (due to EOL)](https://wiki.centos.org/About/Product), but will need to adopt another distro that provides us with an old enough glibc version such that we can offer broad compatibility.
+
+For .NET 9, we'll likely drop support for Windows 10 and instead target the Windows 11 CRT.
```
Does this mean that .NET will "likely drop support for Windows 10" almost a full year before Windows 10's end-of-life date (November 2024 vs October 2025)?

tmds commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,221 @@
+
+**NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and  ...
+
+**Runtime and host packs** -- The .NET SDK doesn't include all assets that are required for all scenarios, but instead relies on many assets being available on a NuGet feed (nuget.org or otherwise). That's a good design point for a variety of reasons. Runtime packs are a concrete scenario, which are used for building self-contained apps. Red Hat should be able to (A) build RHEL-specific runtime p ...
```
Can you include the RHEL-specific rid explicitly in this section (as it gets used in the rest of the doc)?
Something like:

```suggestion
**Runtime and host packs** -- The .NET SDK doesn't include all assets that are required for all scenarios, but instead relies on many assets being available on a NuGet feed (nuget.org or otherwise). That's a good design point for a variety of reasons. Runtime packs are a concrete scenario, which are used for building self-contained apps. Red Hat should be able to (A) build RHEL-specific runtime packages, (B) publish them to a feed of their choice, and (C) enable their users to download them by specifying a RHEL-specific RID (`rhel-x64`) as part of a typical .NET CLI workflow. RHEL-specific packs would not be compatible in the same way as `linux-x64`. They packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes. For example, the RHEL-specific runtime for RHEL 8 would only support the default OpenSSL version that is provided in the RHEL 8 package feed.
```

richlander replied:
Your edit makes it sound like you would not support `linux-x64` NuGet assets. That doesn't match my intuition. Can you elaborate on your thinking?

tmds replied:
I meant to clarify that the packs built by Red Hat have the `rhel-x64` rid, and not the `linux-x64` rid.

> RHEL-specific packs would not be compatible in the same way as `linux-x64`. They packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes.

Maybe this makes it more clear:

> RHEL-specific packs would not be compatible in the same way as `linux-x64` packs. The `rhel-x64` packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes.

tmds replied:
Adding some more text:

> RHEL-specific packs would not be compatible in the same way as the `linux-x64` packs (built by Microsoft). The `rhel-x64` packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes, while the `linux-x64` packs work across a range of distros.

richlander replied:
You are still removing my text on `linux-x64` NuGet assets. I'm talking about a third-party NuGet package. That remains relevant and needs to be explained.

How about this:

> RHEL-specific packs would not be compatible in the same way as the `linux-x64` packs (built by Microsoft). The `rhel-x64` packs would (presumably) only support being run on Red Hat Enterprise Linux family OSes, while the `linux-x64` packs work across a range of distros. However,`linux-x64` NuGet assets (in third-party NuGet packages) would be usable, compatible, and restorable with both Red Hat (distro-specific) and Microsoft (portable) builds of .NET.

tmds replied:
> linux-x64 NuGet assets (in third-party NuGet packages) would be usable, compatible, and restorable with both Red Hat (distro-specific) and Microsoft (portable) builds of .NET.

I think this belongs in the previous paragraph on _NuGet packages_.

richlander replied:
I'm sorry. You were right the whole time. Thanks for being patient.

Check my latest commit.

richlander replied:
https://github.com/dotnet/designs/pull/260/commits/8eb6dae68635228eacfea2dbdc0a430114267ea8

tmds replied:
lgtm, thanks.

tmds commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,221 @@
+
+We'll use Red Hat as an example of an organization that builds .NET from source, to complete this discussion. They have raised concerns on [source-build not playing nicely with `linux-x64`](https://github.com/dotnet/runtime/pull/62942#issuecomment-1056942235).
+
+For Red Hat, it makes sense to accept and support `linux-x64` assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` assets. It's easiest to describe this design-point in terms of concrete scenarios.
```
The rid chosen by a source-build maintainer should not be the same for distribution versions that are not binary compatible. That means source-build rids are likely to have versions in them.

Can you use `rhel.8-x64` instead of `rhel-x64` to make this visible?

richlander replied:
Yes. I realized this after publishing the doc. I'll update this.

richlander replied:
Ughh. I added a mess of an answer. Thoughts?

https://github.com/dotnet/designs/pull/260/commits/840970625f490356d6de38aceea67320987f4fa9

tmds replied:
lgtm.

> However, for that to work, we would need to add `rhel.8-x64` to `runtimes.json`. Otherwise, NuGet won't be able to realize that `rhel.8-x64` should be treated as `linux-x64` for restoring NuGet assets. However, we want to freeze `runtimes.json`. That topic is left as an unresolved issue in this spec, at least for now.

dotnet/runtime has support for patching the `runtime.json` graph with the source-build rid. See https://github.com/dotnet/runtime/pull/50818 by @ericstj.

richlander replied:
Yup. You are right. Will resolve that. Eric separately gave me the same context.

richlander replied:
Added to the doc.

marek-safar commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,221 @@
+## General approach
+
+- Ensure `runtimes.json` is correct.
+- Freeze `runtimes.json`.
```
Freeze completely? How are we going to support a new OSes (e.g. Fuchsia/Zirkon) ?

richlander replied:
They use an existing host, such that the OS uses the `linux-x64` RID. If they have specific needs, then we can discuss it.

teo-tsirpanis commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,221 @@
+|------------|----------------------|
+| unix       | All Unix-based OSes (macOS and Linux), versions, and architecture builds. |
+| win        | All Windows versions and architecture builds. |
+| linux-arm64| All Linux glibc-based distros, for Arm64. |
```
`linux-arm32`? Did you forget it or will it stop being supported? (I don't personally care, just asking)

richlander replied:
Good catch. An oversight.

https://github.com/dotnet/designs/pull/260/commits/e7582b563f3359e6889f6ef07d4e3591073ba45a

eerhardt commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,244 @@
+- `win-x86`
+- `macos-arm64`
+- `linux-arm64`
+- `linux-x64-musl`
```
If this is supposed to be an actual RID (and it feels like it is since the first 3 are RIDs and not triplets), the value is flipped. In .NET RIDs it is: `linux-musl-x64`.

richlander replied:
Sorry. I always write that wrong. Thanks!

```suggestion
- `linux-musl-x64`
```

eerhardt commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,244 @@
+
+## Biggest problems
+
+The following are the biggest problems we see:
```
I believe https://github.com/dotnet/runtime/issues/59803 should be added to this list. The difference between the first bullet below and this issue is that the first bullet below are for new OS names. https://github.com/dotnet/runtime/issues/59803 is for new versions of OSes that we already know about. (And needs servicing changes as well.)

richlander replied:
I added that. Feel free to edit if I messed up the description.

eerhardt commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,244 @@
+- This environment, RID triplet (OS-CRuntime-arch)
+- This environment, RID single (OS-only).
+
+Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.
```
What if I wanted to create a managed .dll that worked across all `macOS` architectures? The same scenario as the above:

> There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) 

Except for `macOS`-specific. Like P/Invoking into macOS native APIs.

The same scenario can be made for Linux and reading certain files like `/proc/self`, etc.

eerhardt replied:
Or is the expectation that .NET Libraries should be using TFMs (`net6.0-macos`, `net6.0-windows`, etc) for those scenarios?

richlander replied:
My understanding is that the use of macOS-specific code is limited and that we rely primarily on `unix` in our own codebase. As a result, I saw `osx` as something we should wait for more concrete scenarios. If that's wrong, we can add `osx` to the macOS probing list now.

The TFMs are really fully separate. They can be use for the purpose you intend, but not for code that is intended as `net6.0`. For that, you need to use RIDs, as you know.

richlander replied:
I'll add this to the spec as something to resolve. It will mostly go away, since `osx-x64` will eventually age out.

tkapin commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,244 @@
+For .NET 7:
+
+- For Windows, .NET will target the Windows 10 CRT.
+- For Linux with glibc, .NET will target CentOS 7.
```
Should this be codified to a concrete glibc version (e.g., 2.17 that is used on CentOS 7)) rather than CentOS version itself to make this clear for other distros?

richlander replied:
Yes and no. I think that's somewhat confusing the point. My understanding is that we're targeting the glibc that CentOS 7 has, whatever it is, making the specific version irrelevant. Fair?

richlander replied:
A little more context. We target the glibc in CentOS 7 to build our `linux-x64` product. If there are distros with an older glibc, then we don't support them, by definition. If .NET is run on a distro with a same/newer glibc, then it is all good. If you build .NET yourself on a distro with a different glibc version, that doesn't matter. It has nothing to do with this plan.

jkotas replied:
I agree with @tkapin that this should be speced using the glibc version. A few reasons:
- If I have some random Linux distro with given glibc version, I should be able to tell whether it is compatible without downloading CentOS 7 just to find the glibc version that happens to be in it. We should not take favorites with the distros.
- The reference distro used by Microsoft official build is likely going to change over time.

It is fine for the spec to explain the underlying reason behind the choice (e.g. that it is the version in CentOS that is the oldest among supported mainstream Linux distros currently).

This applies for both glibc and musl.

tmds replied:
_RHEL/CentOS 7 is so old, there can't be anything older._ :smile: 

It has glibc v2.17.

```
$ podman run centos:7 yum list installed glibc
Loaded plugins: fastestmirror, ovl
Installed Packages
glibc.x86_64                        2.17-317.el7                         @CentOS
```

tkapin replied:
To add one more on top of @jkotas reasons, if we expressed this dependency in terms of distro version, that distro would either have to guarantee that it doesn't offer glibc upgrade to another version within the same version of the distro, or we would have to ensure that we support the earliest version of glibc shipps with that distro version.

We should definitely list the OSes that are supposed to be compatible, but since the primary thing is compatibility with glibc, I think it should be canonically expressed via glibc version with the distro version deduced from it rather than the other way around.

I'm not having an extremely strong opinion about this though so if you feel the former would be more transparent, I'm ok with that.

richlander replied:
Fixed. You guys are right.

tmds commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,244 @@
+
+For Red Hat, it makes sense to accept and support `linux-x64` assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` assets. It's easiest to describe this design-point in terms of concrete scenarios.
+
+**NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and  ...
```
> As a result, it makes sense for Red Hat to support `linux-x64`. There is no technical downside to Red Hat doing so.

Red Hat doesn't do anything to support `linux-x64`. It is people publishing `linux-x64` assets (like Microsoft) that declare compatibility with a range of glibc based Linux distros (like RHEL).

richlander replied:
I think we're talking past one another.

The question is:

- When a dev using the RHEL variant of .NET restores a NuGet asset with a `linux-x64` as an asset (and no other Linux compatible asset), will they get it?

The `linux-x64` pack assets that Microsoft publishes are very different than the ones that the community publishes in NuGet packages.

tmds replied:
Maybe use RHEL instead of Red Hat.

Using Red Hat makes it sound like we do the work. We don't.
The `SkiaSharp` NuGet authors are the ones doing the work by providing native assets that (should) work across the `linux-x64` distros, including RHEL.

richlander replied:
Fixed that.

omajid commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,240 @@
+
+- Ensure `runtimes.json` is correct.
+- Freeze `runtimes.json`.
+- Continue to use `runtimes.json` for all NuGet scenarios.
```
Do you have any pointers to what sort of things are covered by nuget scenarios?

richlander replied:
I don't. Someone else might, but it is really two parts:

- Folder names that are using in NuGet packages.
- The algorithm used by `dotnet restore` in matching the host or user-supplied RID to NuGet package folders.

Fair?

omajid commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,240 @@
+For .NET 7:
+
+- For Windows, .NET will target the Windows 10 CRT.
+- For Linux with glibc, .NET will target CentOS 7.
```
Is that the same for all architectures? For example, in addition to x64, will the CentOS 7 version of glibc also be the baseline for `arm64` and `s390x` architectures?

richlander replied:
That would be my base assumption.

tmds replied:
They can be different because they are anyhow incompatible.

If you don't need to support CentOS 7 on an architecture, there may be build time benefits to using something newer.

For s390x, Microsoft isn't building any `linux-s390` artifacts for .NET, and the Red Hat artifacts will use a source-build rid (like `rhel.8-s390x`).
The glibc baseline serves for NuGet package authors that want build binaries that work across a range of distros. The basline should be the lowest version of CentOS/RHEL that supports s390x.

richlander replied:
I added some text on this.

omajid commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,240 @@
+- This system supports this RID, per the .NET host.
+- This code will run on / requires a system that supports this RID, per the author.
+
+The following table describes RIDs supported by the .NET host. It is exhaustive, per the hosts provided by Microsoft.
```
How do we deal with architectures like `s390x` that Microsoft doesn't support and/or provide builds for?

richlander replied:
That's what this is about: https://github.com/dotnet/designs/pull/260/files#diff-20aa11d3c5c8d54fc1c45f4b81927255d3951fdd7cf58589fc6d03a2dd2a6a1bR221

The proposal is to NOT add any new RIDs, but enable source-builds to add additive `runtimes.json` files that augment the graph. So, frozen is kind a soft term. It means that we want to avoid creating any new official RIDs, but that distro maintainers can create their own for their own distro. The assumption is that they are only needed to create a link between the user and the runtime/host packs and not NuGet. As a result, these RIDs don't need to be stored anywhere else. We'll accept a file like `runtimes-s390x.json` in upstream. It would probably be <20 lines.

Fair?

omajid replied:
> The assumption is that they are only needed to create a link between the user and the runtime/host packs and not NuGet

In the SkiSharp scenario, the authors of SkiaSharp might want to publish `linux-s390x` assets on nuget.org. What would that look like?

Alternatively, do you think it might be worth it to blanket-add all the current architectures supported by mono to runtime.json before freezing it?

tmds replied:
> In the SkiSharp scenario, the authors of SkiaSharp might want to publish linux-s390x assets on nuget.org. What would that look like?

They can add an so file in their NuGet package at runtimes/linux-s390x, and then the package will support s390x.

```
$ wget --no-check-certificate https://www.nuget.org/api/v2/package/SkiaSharp.NativeAssets.Linux/2.88.0-preview.232
$ unzip 2.88.0-preview.232
$ cd runtimes/
$ tree
.
├── linux-arm
│   └── native
│       └── libSkiaSharp.so
├── linux-arm64
│   └── native
│       └── libSkiaSharp.so
├── linux-musl-x64
│   └── native
│       └── libSkiaSharp.so
└── linux-x64
    └── native
        └── libSkiaSharp.so
```

The `linux-s390x` rid could be added as part of source-build which means it will be usable on the resulting .NET distribution.

The reason for adding it to upstream `runtimes.json` is to make it a known rid for to **all** .NET distributions. Then they can publish against that rid. Otherwise it would be unknown.

richlander replied:
That's a good point. Agree. Is that good @omajid? If so, I'll update with the spec with the wording from @tmds.

richlander replied:
Updated the doc.

omajid commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,249 @@
+|------------|----------------------|
+| unix       | All Unix-based OSes (macOS and Linux), versions, and architecture builds. |
+| win        | All Windows versions and architecture builds. |
+| linux-arm32| All Linux glibc-based distros, for Arm32. |
```
This should be [`linux-arm`](https://github.com/TobiasLaving/runtime/blob/312c66f1fc2f749f56612999cb1adab9ca7fde59/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json#L1460), right?

richlander replied:
```suggestion
| linux-arm| All Linux glibc-based distros, for Arm32. |
```

omajid commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,249 @@
+|------------|----------------------|
+| unix       | All Unix-based OSes (macOS and Linux), versions, and architecture builds. |
+| win        | All Windows versions and architecture builds. |
+| linux-arm32| All Linux glibc-based distros, for Arm32. |
```
Are there any plans to add [`linux-armv6`](https://github.com/dotnet/runtime/pull/62594) here?

richlander replied:
Ya. I'll add it. We can decide whether it fully makes sense later.

bording commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+
+For Red Hat, it makes sense to accept and support `linux-x64` NuGet assets, but not to produce or offer them. Instead, Red Hat would want to produce `rhel-x64` runtime pack assets. It's easiest to describe this design-point in terms of concrete scenarios.
+
+**NuGet packages** -- NuGet authors will typically value breadth and efficiency. For example, a package like `SkiaSharp` might target `linux-x64`, but not `rhel-x64`. There is no technical reason for that package to include distro-specific assets. Also, if the author produced a `rhel-x64` asset, they would need to consider producing an `ubuntu-x64` and other similar assets for other distros, and  ...
```
I feel like this document generally, and this section specifically, is making some assumptions that I don't think are justified. There are absolutely scenarios where distro-specific assets need to be shipped in a NuGet package.

As an example of this, take a look at https://www.nuget.org/packages/LibGit2Sharp.NativeBinaries/2.0.306

LibGit2Sharp is a managed wrapper around the native libgit2 library and has to accommodate the native dependencies of libgit2. One of the dependencies is OpenSSL, which varies wildly among all the distros that .NET supports.

In order to have a reasonable chance of having a native binary that just works on all platforms, I have had to spend a lot of time understanding the RID graph and shipping enough distro-specific binaries to cover all the supported Linux distros

I will point out that libgit2 has recently changed how they are binding to OpenSSL which has let me simplify things down to more of an "ideal" situation (see the [newer package](https://www.nuget.org/packages/LibGit2Sharp.NativeBinaries/2.0.315)), but that was largely out of my control since I'm not a libgit2 maintainer.

If the ability to ship distro-specific assets in a NuGet package and have the proper one selected had not been a feature that I could count on, then I would not have been able to support LibGit2Sharp on Linux at all for the past several years. Maybe I'm misunderstanding something, but it seems like this proposal is removing this feature.

omajid replied:
Mind you, https://github.com/libgit2/libgit2sharp/pull/1714 changed the model used by libgit2sharp. AFAIK, the change basically makes libgit2sharp try and load all native libgit2 libraries one by one until one of them can be loaded successfully (ie, links to a version of OpenSSL available on the system). The actual name (and RID) is not too relevant in this scenario.

I feel like your use-case would still work if nuget did not "know" about distro-specific RIDs.

richlander replied:
This is great feedback. We discussed it.

There are three models:

- Don't support specific distros as a general concept (the current proposal).
- Fully support specific distros as a general concept (some variation on status-quo).
- Adopt a model that is a variation on the Python manylinux wheel plan, including support for OpenSSL.

I like what @omajid is proposing. It's the same approach we use for OpenSSL. We don't use RIDs. Also, the current proposal works for the vast majority of cases and is a massive simplification. I'm hoping we can keep it.

tmds replied:
libgit2sharp was suffering from the rid scalability issue: it was unable to deal with unknown rids.

Using `NativeLibrary.SetDllImportResolver` it now tries the different `so` files. The code is here: https://github.com/libgit2/libgit2sharp/blob/97bee65fd296f1c7dd2d1d64581c170f45b584e1/LibGit2Sharp/Core/NativeMethods.cs#L99-L121

This way, the library can have its own logic for picking `so` files. It's not limited to/by rids.

bording replied:
Yes, this logic was added, but I consider it a fallback solution, and not something desirable to rely on as the primary logic, for a couple of reasons:
- It only works if the entire runtimes folder is shipped as part of the application, which is only going to be true if you use one of the cross-platform publishing options
- Having to attempt to load every binary until you find one that works seems like it would have a negative impact on startup perf.

If libraries no longer can assume that the best native asset is selected for them by the framework/NuGet, then that means they have to manually ensure that all binaries are included in the publish output to be able to select from them at runtime. That can pretty massively increase the size of the output. Using LibGit2Sharp.NativeBinaries 2.0.306 as an example, that means going from having a single ~1MB .so file in your output to having 9 different copies, and that's assuming I'd have some way to know to only copy the Linux binaries and not the entire runtimes folder.

tmds replied:
> It only works if the entire runtimes folder is shipped as part of the application, which is only going to be true if you use one of the cross-platform publishing options

This way of trimming it isn't scalable as you need to build and include an `so` file for every platform you want to trim.

A scalable solution is to allow to control the trimming more specifically. For example add a property to csproj that enables the user to choose the supported OpenSSL versions, and include `so` files based on that.

> Having to attempt to load every binary until you find one that works seems like it would have a negative impact on startup perf.

I doubt this has a measurable impact.

> Using LibGit2Sharp.NativeBinaries 2.0.306 as an example, that means going from having a single ~1MB .so file in your output to having 9 different copies

It seems the latest version doesn't have distro specific versions. Why did it change?

richlander replied:
> perf

.NET probes for OpenSSL. No one has ever submitted a perf report on that.

@bartonjs -- can you point to the probing code for OpenSSL?

jkotas replied:
> can you point to the probing code for OpenSSL?

OpenSSL probing code in dotnet/runtime: https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Security.Cryptography.Native/opensslshim.c#L54-L121

OpenSSL probing in libgit2: 
https://github.com/libgit2/libgit2/blob/82f526a9129373d899782c411774c8e396cae4e4/deps/ntlmclient/crypt_openssl.c#L63-L69 . Notice that it does not have support for probing for OpenSSL 3 yet so it is going to break once distros switch to OpenSSL 3 only. dotnet/runtime has probing for OpenSSL3 that adds a whole new level of shimming and complexity.

> perf

The probing is relatively expensive when looked at in isolation (it is file I/O). It is relatively cheap compared to e.g. all JITing that happens during startup of a typical .NET app.

> I see there is a linux-x64 asset. What's the intended behavior of that asset? By definition, it can run on any (glibc) distro.

Yes, this works like bow that since libgit2 added a configuration to probe for OpenSSL (https://github.com/libgit2/libgit2/pull/5974) and the linux-x64 binaries in the nuget package are built with this option on.

bording replied:
> This way of trimming it isn't scalable as you need to build and include an so file for every platform you want to trim.

That's not how publishing works currently, though. It uses the RID graph to select the best option from the assets in the package. My goal was always to have enough coverage so that the supported .NET platforms would just work. That *was* broken sometimes due to the RID graph not being updated...

> It seems the latest version doesn't have distro specific versions. Why did it change?

I mentioned that earlier here: libgit2 has recently added a dynamic binding option for OpenSSL, which has let me simplify the package.

> I see there is a linux-x64 asset. What's the intended behavior of that asset? By definition, it can run on any (glibc) distro.

The set of binaries in the package is a result of analyzing the distros that .NET Core 3.1 supported and the RID graph to ship the fewest number of binaries with the maximum amount coverage. The binary shipped as `linux-x64` was intended to be a fallback that would work when a more specific binary was not chosen. That meant I needed to understand the glibc and OpenSSL versions in each distro to be able to construct the set of binaries I shipped.

This was a large effort, but the end result was that most/all the distros would work out of the box. However, when the RID graph was not updated for a new .NET Core release and a new distro/version incorrectly defaulted back to `linux-x64`, then that fallback would fail. That's where @tmds PR came in and added an additional fallback to try and see if one of the binaries in the package might work even though the RID graph wasn't able to select it.

jkotas replied:
> For example add a property to csproj that enables the user to choose the supported OpenSSL versions, and include so files based on that.

The problem does not end with OpenSSL. The same problem exists for many other libraries.

I believe that the only truly scalable option for Linux ecosystem is to have building from source as an option (https://github.com/NuGet/Home/issues/9631). Building from source is capable of producing a binary that works for the specific configuration of your system without the package maintainer pre-building it for you.

bording replied:
> The problem does not end with OpenSSL. The same problem exists for many other libraries.

Yes, and this was the main point of the feedback I was trying to raise here. The proposal currently assumes that there is never a reason to care about the specific distro you're running on, but that is demonstrably not true.

This all comes down to there being a disconnect between the .NET ecosystem expectations of sharing pre-compiled binaries, and the Linux ecosystem that expects you to share source and compile it on your own system. Shipping a Linux binary in a package is always going to be an uphill battle, so I would like to see the .NET side of things evolve in a way that makes it easier, not harder.

tmds replied:
> Notice that it does not have support for probing for OpenSSL 3 yet so it is going to break once distros switch to OpenSSL 3 only. dotnet/runtime has probing for OpenSSL3 that adds a whole new level of shimming and complexity.

Let's assume OpenSSL 3 won't be supported by a libgit2 shimming (immediately) and you want to solve it issue using distro rids.

Fedora 36 should use the OpenSSL 3 build.

Fedora 37 should be using the OpenSSL 3 build also. That doesn't work automagically today, because the rid grah doesn't describe a supports relationship between Fedora 37 and Fedora 36.

For your package to remain maintainable, this needs to be addressed (and preferably not by bloating the graph but by making the rid logic smarter).

I've only mentioned Fedora here, but OpenSSL will be adopted by Ubuntu, Debian, so you need to manage this problem for them too.

Your package now looks like this:

```
linux-x64 -> supports OpenSSL 1, 2, 3 and picks using `NativeLibrary`
fedora.36 -> OpenSSL 3
ubuntu.22.04 -> OpenSSL 3
debian.12 -> OpenSSL 3
...
```

Note that you are including the distro specific assets solely for the purpose of trimming. They bloat your package and require you to track in what version a distro adopts a new OpenSSL version.

> The proposal currently assumes that there is never a reason to care about the specific distro you're running on

`NativeLibrary` can be used, and should be used to make `linux-x64` work across a range of distros.

What is lost, is the ability to trim against a rid.

tmds replied:
@bording I hope this example helped illustrate how unscalable the rid mechanism is?

And, if it should stay, inheritance between successive distro versions is a must-have, which is now missing.

tmds replied:
> And, if it should stay, inheritance between successive distro versions is a must-have, which is now missing.

To elaborate. The rids as we have them are not scalable. The alternative to the simple model that is proposed here, is to have _semantic rids_. That means rids follow a `<distro>.<version>-<arch>` naming scheme. Information is derived from that instead of hard coding it in the json file (e.g. Fedora 36 -> Fedora 35). The json file should then only contain non-derivable information (like what distros are glibc/musl based).

This is a much more complex solution.

jkotas commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+
+## Minimum CRT/libc version
+
+This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.
```
Using C runtime version to define the OS version makes sense on Linux. It is the best approximation of the OS version to use for determining binary compatibility (of binaries written in C/C++).

Using C runtime version for this purpose makes less sense on Windows and macOS. Windows and macOS have well-defined OS version. Both Windows (Windows SDK) and macOS (XCode) native toolchains are oriented on using the OS version for this purpose ([WINVER](https://docs.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt) on Windows and `-mmacos-version-min` on macOS). We should use OS version for determining binary compatibility on Windows and macOS.

richlander replied:
Do you want me to add that text, or adapt its meaning into the current text?

jkotas replied:
I think adapting the meaning into the current text would look better. I would lead with minimum OS version as a documented contract, per .NET version. libc version is Linux-specific approximation of OS version across distros for the purposes of this contract.

am11 commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+
+## Context
+
+RIDs are (for the most part) modelled as a graph of [target triplet](https://wiki.osdev.org/Target_Triplet) symbols that describe legal combinations of operating system, chip architecture, and C-runtime, including an extensive fallback scheme. This graph is codified in [`runtime.json`](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json), which i ...
```
> target triplet

AFAIK, target triplets have more cons than pros. It really depends on target audience, but there are negative connotations with the concept of `triplet` in native tooling communities. They consider it vague and archaic concept. e.g. for the modern ISA, the compiler toolchain [documentation](https://www.sifive.com/blog/all-aboard-part-1-compiler-args) was written in terms of ABI (`-mabi`), architecture (`-march`) and microarchitecture (`-mtune`). In case of .NET tooling, `--arch` and `--os` are  relatable concepts.

richlander replied:
`--arch` and `--os` are indeed nicer to work with. They came out of our effort to make it easier to support emulated environments (like Rosetta 2 on macOS). They are however just a convenience wrapper over the RID concept. More importantly, NuGet packages need a currency for describing assets intended for various. The triplets are the simplest scheme we've found for that. We haven't found a better system.

am11 replied:
I completely understand and it is probably too late to do anything about it. Since RID is inspired by triplets, I just wanted to point out that modern toolchains prefer explicitly specifying each aspect of platform individually over having a triplet scheme to interpret magic strings, which parser/consumer would always need to remember. They also started their journey with triplets but over the years, found it as an unnecessary overhead.

richlander replied:
That all makes sense when you are using a compiler to produce artifacts that you will use. I fail to see how that works for a package manager that supports a broad computing ecosystem. Those specific flags also seem too low-level. By definition, they are not intended (per my read) to be used outside of the RISC-V ecosystem.

am11 replied:
> I fail to see how that works for a package manager that supports a broad computing ecosystem.

`rid: linux-x64` versus `arch: x64, os: linux`

richlander replied:
OK. You literally just mean using attributes. Got it. Yes, that would have been a different initial design.

Package folks would then look like:

AttributeSpecificAssets
  A
    attributes.json
    Foo.dll
  B
    attributes.json
    Foo.dylib

I'm sure it wouldn't look exactly like that, but that's the basic idea. It could work. A nice characteristic of the current design is that the folder names (for A and B) encode the target so the lookup mechanism is easier. Other things (as you've called out) with that design are worse.

am11 replied:
Yup, the parsing looks like this: https://github.com/dotnet/runtime/blob/b42adaddff9047e32c9d7f4145746dbb579758f2/src/libraries/Microsoft.NETCore.Platforms/src/RID.cs#L59 and we do that in msbuild scripts and C/C++ separately. :)

jkotas replied:
> AFAIK, target triplets have more cons than pros. It really depends on target audience, but there are negative connotations with the concept of triplet in native tooling communities.

I agree that the classic `machine-vendor-operatingsystem` triplets have archaic feel, due to the `vendor` field and to all sorts of cryptic suffixes that tend to be appended to them.

Most RIDs are `operatingsystem-machine` that does not the same feel. Maybe the doc should not lead with explaining RIDs as target triplets to avoid giving readers wrong impression.

Python `Platform Tag` defined in [PEP 0425[(https://peps.python.org/pep-0425/) is much more similar to RIDs than target triplets. The definition of the Python `Platform Tag` does not make any associations with target triplets.

agocke replied:
Target triplets do indeed have problems and the parsing is annoying. But I think the fundamental problem is the same -- in order to provide binary compatibility between components, we need to have an encoding for whatever's necessary for binary compatiblity on that platform.

For Mac and Windows, OS version and architecture is probably sufficient. For Linux, we'll also need libc type (and version, although it could be implicitly tied to the TFM).

jkotas replied:
Yes, we are in agreement about what we want to do. This is just a discussion about wording - whether it is helpful to describe RID as target triplet or not.

am11 replied:
Yes, RID is essentially a close-enough hint and we cannot reliably capture the "desired system configuration" with it accurately. As you pointed out in other thread @agocke, that was not the goal RID was designed to achieve. We perform run time introspections and we have abstraction layers in place to adapt to the system we are running on; in (a fex/countable) places where it makes a difference in dotnet/runtime. This is probably the extent to which 3P package, with OS specific native artifacts,  rely on RID.

richlander replied:
I am happy to remove the linkage with target triplet if that is what folks want.

am11 commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+- `linux-x64`
+- `unix`
+
+A source-built RID (here `rhel.8-x64`) will typically be distro-specific and versioned. Red Hat would naturally want to build .NET specifically and separately for RHEL versions, like RHEL7 and RHEL 8, since the packages available in each of their versioned Red Hat package feeds will be different. There might be other differences to account for as well. An unversioned RID (like `rhel-x64`) would n ...
```
> Red Hat would naturally want to build .NET specifically and separately for RHEL versions,

Do we know if other programming platforms (Go, Python, Node.js etc.) also use RH specific triplets? If they do not (which I think is the case), then why do we see it natural requirement for .NET, specifically?

tmds replied:
Go is well known for its ability to to produce native executables for different platforms.
It has no dependency on a c-library, instead it has its own Go-implementation that makes syscalls directly.
Consequently it doesn't have the need to describe this dependency as a rid.

I don't know the story for building native executables from Python/JavaScript.

terrajobst commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+
+- Ensure `runtimes.json` is correct.
+- Freeze `runtimes.json`.
+- New RIDs will only be added for [interchange across multiple platforms](https://github.com/dotnet/designs/pull/260#discussion_r843009872), which isn't expected.
```
When you say

> interchange across multiple platforms

You really mean "between different distributions of .NET", right?

richlander replied:
```suggestion
- New RIDs will only be added for [interchange between different distributions of .NET](https://github.com/dotnet/designs/pull/260#discussion_r843009872), which isn't expected.
```

terrajobst commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+
+The new scheme will be based on the following:
+
+- Each host is built with a hard-coded set of RIDs that it probes for, both triplets and singles.
```
Does this mean that we either

(1) Hard code the (large set) fallbacks into the host
(2) Evaluate the fallbacks at build time and put them in the (small) set of RID specific folders the host is built for?

richlander replied:
Neither. This is covered later in the doc. Each host will know about 2-3 RIDs and none other. If that doesn't work, there will be a fall-back mechanism (which will just be a MSBuild + `runtimeconfig.json` property) to use the RID catalog. Make sense?

terrajobst commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,252 @@
+
+Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.
+
+More generally, the host will probe for:
+
+- This environment, RID triplet (OS-CRuntime-arch)
+- This environment, RID single (OS-only).
```
If we build a host for a specific RID, why do we need to do any probing at runtime? Asked differently, why can't we do all the probing at build time?

richlander replied:
Because of portable apps.

bording commented:
> I think the main impact of the design is that package maintainers must change from using distro-rids for separating native libraries to group them under a portable rid (e.g. linux-x64) and use `NativeLibrary` to load the appropriate native library.

The problem I see with saying package maintainers have to use this approach is that it is overlooking how publish output changes if you specify a rid or not.

A portable publish will copy all of the rid folders into the output, but a publishing with `-r linux-x64` will copy just the content of the package's `linux-x64` folder into the main publish output folder.

That makes writing code that can use `NativeLibrary` to search and find the right binary much more complicated because you can't always assume there is a `linux-64` folder in the project output to be searching in.

tmds commented:
My comment was to highlight the complexity that gets pushed towards the package maintainers.

Note that if your library has different native libraries for glibc Linux distros, in order to make `-r linux-x64` portable (that is: work across a that range of distros), you already need to use `NativeLibrary` to select the appropriate native library.

With this proposal, in the case the user doesn't specify a rid, the library now also needs to use `NativeLibrary` because host will no longer do the distro rid-graph based selection.

bording commented:
> My comment was to highlight the complexity that gets pushed towards the package maintainers.

:+1: And that is what I've been somewhat trying to push back on through my comments here. Or at least acknowledgement that this is understood and documented appropriately on how to author packages in this more complicated way.

> Note that if your library has different native libraries for glibc Linux distros, in order to make -r linux-x64 portable (that is: work across that range of distros), you already need to use NativeLibrary to select the appropriate native library.

> With this proposal, in the case the user doesn't specify a rid, the library now also needs to use NativeLibrary because host will no longer do the distro rid-graph based selection.

And my previous comment was trying to point out that writing this `NativeLibrary` code is much harder with this proposal because of needing to handle two different cases, which I don't think I've seen acknowledged/understood on this PR yet.

tmds commented:
> And my previous comment was trying to point out that writing this NativeLibrary code is much harder with this proposal because of needing to handle two different cases, which I don't think I've seen acknowledged/understood on this PR yet.

It depends.

If you know the names of the libraries, you can pass them to `NativeLibrary` and it will probe the paths that have the native libraries. Then the code is the same.

If you want to iterate yourself the directories being searched to discover at runtime what native libraries are available (e.g. based on a naming pattern), then there is no way to get these directories. They could be exposed through a property on `NativeLibrary` for example.

bording commented:
> It depends.

> If you know the names of the libraries, you can pass them to NativeLibrary and it will probe the paths that have the native libraries. Then the code is the same.

> If you want to iterate yourself the directories being searched to discover at runtime what native libraries are available (e.g. based on a naming pattern), then there is no way to get these directories. They could be exposed through a property on NativeLibrary for example.

To make sure everyone is talking about the same thing, I want to be clear and describe the scenario. We are talking about a NuGet package that has a distro-specific native dependency.

Currently, the way to do that is to ship a package with as many distro-specific RIDs as needed. For example, the runtimes folder might look something like:
```
+---alpine-x64
|   \---native
|           libgit2-106a5f2.so
|
+---alpine.3.9-x64
|   \---native
|           libgit2-106a5f2.so
|
+---debian-arm64
|   \---native
|           libgit2-106a5f2.so
|
+---debian.9-x64
|   \---native
|           libgit2-106a5f2.so
|
+---fedora-x64
|   \---native
|           libgit2-106a5f2.so
|
+---linux-x64
|   \---native
|           libgit2-106a5f2.so
|
+---osx
|   \---native
|           libgit2-106a5f2.dylib
|
+---rhel-x64
|   \---native
|           libgit2-106a5f2.so
|
+---ubuntu.16.04-arm64
|   \---native
|           libgit2-106a5f2.so
|
+---ubuntu.18.04-x64
|   \---native
|           libgit2-106a5f2.so
|
+---win-x64
|   \---native
|           git2-106a5f2.dll
|           git2-106a5f2.pdb
|
\---win-x86
    \---native
            git2-106a5f2.dll
            git2-106a5f2.pdb
```

This is tricky to get right and does require ongoing work to ensure the binaries shipped are comprehensive. It also requires that the RID graph is maintained and accurate.

---
The proposed alternative here seems to be something like the following instead:
```
+---linux-x64
|   \---native
|       +---alpine-x64
|       |       libgit2-106a5f2.so
|       |
|       +---debian-x64
|       |       libgit2-106a5f2.so
|       |
|       +---fedora-x64
|       |       libgit2-106a5f2.so
|       |
|       +---rhel-x64
|       |       libgit2-106a5f2.so
|       |
|       \---ubuntu-x64
|               libgit2-106a5f2.so
|
+---osx
|   \---native
|           libgit2-106a5f2.dylib
|
+---win-x64
|   \---native
|           git2-106a5f2.dll
|           git2-106a5f2.pdb
|
\---win-x86
    \---native
            git2-106a5f2.dll
            git2-106a5f2.pdb
```
With this sort of package layout, it is now on the author to write code that can search inside the `runtimes\linux-x64` folder and try to load each binary with `NativeLibrary` to find one that works. This sort of code would vaguely look something like [this](https://github.com/libgit2/libgit2sharp/blob/1af49546917b69499f2d41cf844509c53b37eaca/LibGit2Sharp/Core/NativeMethods.cs#L99-L121).

As things currently stand, I see some problems with this proposal, and it actually isn't possible to author a package like this with the way NuGet works right now.

Problems:
1. Differences between portable and RID-specific publish output

When you build a project or publish it without a RID specified (`dotnet publish`), the publish output gets a copy of the package's entire `runtimes` folder. Code written to use `NativeLibrary` will need to expect this folder structure when searching for the correct Linux binary to load, for example subdirectories of `runtimes\linux-x64\native`.

However, as soon as you publish with a RID specified, the folder structure is collapsed and the contents of the native folder are now top-level in the publish output. Using the proposed folder structure above, with something like `dotnet publish -r win-x64`, that results in just the `git2-106a5f2.dll` file being in the publish output.

What happens with `dotnet publish -r linux-x64`? You might assume that it would copy the contents of the `native` folder and maintain the folder structure defined in the package.

If this was true, how are you supposed to author the `NativeLibrary` loading code that in one case needs to look for binaries in `runtimes\linux-x64\native` and in another instance needs to be looking for files\folders in the root directory? Is there some way at runtime to know if you're running from RID-specific published content or not? The only thing I can think of at the moment would be to check and see if a `runtimes` folder exists or not, but that also has an ongoing maintenance burden of having to maintain the list of distro folders might be in the application root to search.

However, NuGet does not actually seem to work this way currently.

2. NuGet won't honor the proposed folder structure

When you publish with `-r linux-x64`, instead of just copying the contents of `runtimes\linux-x64\native` into the publish output, you get an error:
```
error NETSDK1152: Found multiple publish output files with the same relative path
``` 
It is complaining about seeing more than one copy of the binary as if it's trying to put all of them into the publish folder instead of maintaining the defined folder structure.

If you instead try to put the distro subfolders directly under `runtimes\linux-x64` and not inside the `native` folder, that makes things worse. NuGet now completely ignores the `linux-x64` folder and it doesn't show up inside publish output at all.

You could potentially work around the NETSDK1152 error by naming every single binary differently, but that's not always under your control. It also doesn't solve the `NativeLibrary` problem, and in some ways seems to make it worse.

---
Ultimately, my goal with all of this is to ensure that whatever change ends up happening is done with an awareness that work needs to be done to continue to support packages that need to have distro-specific native binaries. The burden can't be shifted entirely to package maintainers, and in fact the way things currently work make that nearly impossible.

jkotas commented:
> The proposed alternative here seems to be something like the following instead:

The proposed alternative here is the following structure instead: 
```
+---linux-x64
|   \---native
|       +---linux-musl-x64
|       |       libgit2-106a5f2.so
|       |
|       +---libgit2-106a5f2.so
|
+---osx
|   \---native
|           libgit2-106a5f2.dylib
|
+---win-x64
|   \---native
|           git2-106a5f2.dll
|           git2-106a5f2.pdb
|
\---win-x86
    \---native
            git2-106a5f2.dll
            git2-106a5f2.pdb
```

where `libgit2-106a5f2.so` is built in prescribed environment with specific glibc version so that the binary runs on all Linux versions supported by the .NET runtime that the nuget package is targeting. It is a direct equivalent of the approach used by Python to solve this problem: https://peps.python.org/pep-0600/ .

bording commented:
@jkotas I don't see how that is meaningfully different from what I had in my post. And it doesn't solve any of the problems I described. 

This isn't just about glibc vs musl. This is about any sort of native dependency that is different per distro. For example, OpenSSL.

jkotas commented:
The assumption in this proposal is that it is just about glibc vs. musl for the vast majority of native libraries bundled into packages. The Python solution is based on the same assumption.

For OpenSSL 2 vs. 3 problem, you can build one version of the .so that dynamically binds to OpenSSL2 or 3. It is what .NET runtime itself is doing. libgit2 mirrored that solution as well (https://github.com/libgit2/libgit2/blob/main/src/libgit2/streams/openssl_dynamic.c). Or you can build two version of the so like libgit2_openssl2.so and libgit2_openssl3.so and implement a loader that loads the right one at runtime.

If you have a complex library with a many different dependencies, you can do a custom loader that picks up the right flavor for given distro. Yes, it is complex. The assumption is that only a small fraction of packages would need to do something like this.

jkotas commented:
> Are there some numbers that tell us how many packages on nuget.org have native libraries for non-portable rids?

Yes, that would be useful data to have, together with the usage numbers.

I believe that people tend to stay away from packages with native dependencies on Linux today since they are broken too often. For example, we had to drop the dependency on LibGit2Sharp in source link (https://github.com/dotnet/sourcelink/pull/288) and reimplement it in C#. One of the reasons was that the distro-specific libgit2 builds caused too many problems.

trungnt2910 commented:
This issue seems to be focused on relieving of the complexity of having to manage various Linux distros and their versions.

What about other OSes that are completely different from Linux in ways that are not only related to native library ABI, but also managed system library code?

https://github.com/dotnet/runtime/pull/90695, which seems to be related to this issue, is currently breaking managed library builds for Haiku, since the official .NET SDK is not aware of Haiku and the build is now configured to ignore the `runtimes.json` in the repository.

Unlike different Linux distros like Ubuntu and Alpine, which can use the same IL code as any `linux-x64` in most cases, Haiku is as different from Linux as FreeBSD is when it comes to managed libraries, so wouldn't it make sense to still allow new RIDs for new OS support, but not for new Linux distros?

agocke commented:
The intent of this proposal is to move away from OS flavor- and version-specific targeting. This is most notable for Linux because there are so many versions and flavors, but it's equally true for Windows and Mac, where you will no longer be able to target versions individually.

However, different OSes are still classified as different portable RIDs. A portable RID is essentially `<baseOS>[-optional-libc]-arch`. Since Haiku is a different base OS, it would be eligible for its own. It would just take a community member or members willing to port the runtime to Haiku and maintain compatibility. Since it is not one of the officially supported OSes, Microsoft would not provide support directly. This is similar to the status of FreeBSD.

trungnt2910 commented:
> It would just take a community member or members willing to port the runtime to Haiku and maintain compatibility.

https://github.com/dotnet/runtime/pull/86391#discussion_r1247869065

It is being done here, but other members of dotnet is questioning whether a RID for Haiku should be added.

> Since Haiku is a different base OS, it would be eligible for its own.

This means unlike what https://github.com/dotnet/runtime/pull/90695 states, the list of RIDs should not be frozen for good, it should just be less volatile since version numbers and distro flavors are not being updated anymore?

richlander commented:
@elinor-fung @agocke @jkotas I want to merge this content after (almost) four years. I went through and fixed some minor issues, updated dead links, and added a tiny bit of context. The intent was to leave the spec in its 4-year-old state with minimal cleanup. PTAL (can be cursory).

agocke commented on accepted/2022/simplify-rid-model.md:
```diff
@@ -0,0 +1,250 @@
+Examples of triplets today are:
+
+- `win-x86`
+- `macos-arm64`
```
```suggestion
- `osx-arm64`
```
